src/main/java/org/candlepin/subscriptions/product/MidProduct.java

```diff
@@ -0,0 +1,376 @@
+   * For now though, this simplification fits with how we present offerings to the user.
+   * See:
+   * https://docs.google.com/document/d/1t5OlyWanEpwXOA7ysPKuZW61cvYnIScwRMl--hmajXY/edit#heading=h.3aq1apsnbb0o
+   *
+   * <p>Example: SKU RH00001, child SKU SVCRH00001, have attr DERIVED_SKU = "RH00049".
+   * Neither RH00001 or SVCRH00001 have engOids, but RH00049's child SKU SVCRH00049 does.
+   */
+  private MidProduct fetchAndAddDerivedTreeIfExist(ProductService productService) {
+    String derivedSku = attrs.get(Attr.DERIVED_SKU);
+    if (derivedSku != null) {
+      try {
+        // derived SKUs are marketing SKUs, so need to get its service SKUs too and add
+        // the SKUs to the list of children so the engOids are fetched.
+        Optional<MidProduct> mid = productService.getTree(derivedSku).map(this::addDerivedTree);
+        if (mid.isEmpty()) {
+          LOGGER.warn(
+              "No tree found for derivedSku=\"{}\" of offering sku=\"{}\"", derivedSku, sku);
+        }
+      } catch (ApiException e) {
+        throw new RuntimeException(
+            "Unable to retrieve derivedSku=\"" + derivedSku + "\" for offering sku=\"" + sku + "\"",
+            e);
+      }
+    }
+    return this;
+  }
+
+  private MidProduct fetchAndAddEngProdsIfExist(ProductService productService) {
+    Set<String> allSkus = allSkus();
+    LOGGER.debug("Retrieving engOids for skus=\"{}\" of offering sku=\"{}\"", allSkus, sku);
+    try {
+      Map<String, List<EngineeringProduct>> engProds =
+          productService.getEngineeringProductsForSkus(allSkus);
+      addEngProds(engProds);
+    } catch (ApiException e) {
+      throw new RuntimeException(
+          "Unable to retrieve engOids of skus=\"" + allSkus + "\" for offering sku=" + sku, e);
+    }
+
+    return this;
+  }
+
+  private MidProduct addDerivedTree(RESTProductTree derivedTree) {
+    derivedTree.getProducts().stream()
+        .map(MidProduct::createFromDerivedProduct)
+        .forEach(this::merge);
+    return this;
+  }
+
+  private void addEngProds(Map<String, List<EngineeringProduct>> engProds) {
+    engProds.values().stream()
+        .flatMap(engProdList -> engProdList.stream().map(engProd -> engProd.getOid()))
+        .forEach(engOids::add);
+  }
+
+  /**
+   * Get all SKUs involved in the offering, including the offering SKU, child SKUs, and derived SKU.
+   *
+   * @return collection of all SKUs, or empty if none.
+   */
+  private Set<String> allSkus() {
+    var skus = new HashSet<>(children);
+    skus.add(sku);
+    return skus;
+  }
+
+  private void calcCoresAndSocketsForOffering(Offering offering) {
+    String ifl = attrs.get(Attr.IFL);
+    // ifl takes precedence over cores, so don't set cores if ifl exists.
+    String cores = (ifl == null) ? attrs.get(Attr.CORES) : null;
+    String sockets = attrs.get(Attr.SOCKET_LIMIT);
+
+    String virtIfl = attrs.get(Attr.__X_VIRTUAL_IFL);
```

Comment:
I could not find any real-world examples of a derived SKU with an IFL attr. For the sake of symmetry I added the IFL logic to both the physical and virtual cores calculations, even though in reality I could only find examples of physical cores being affected.

Therefore, if this is an example of YAGNI then I can remove it.